File: stripe/src/test/java/com/stripe/android/model/CustomerTest.java
package com.stripe.android.model;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import static com.stripe.android.model.CardTest.JSON_CARD_USD;
import static com.stripe.android.model.CustomerSourceTest.JSON_APPLE_PAY_CARD;
import static com.stripe.android.model.SourceTest.EXAMPLE_ALIPAY_SOURCE;
import static com.stripe.android.model.SourceTest.EXAMPLE_JSON_SOURCE_WITHOUT_NULLS;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

/**
 * Test class for {@link Customer} model object.
 */
public class CustomerTest {

    private static final String NON_CUSTOMER_OBJECT =
            "{\n" +
                    "    \"object\": \"not_a_customer\",\n" +
                    "    \"has_more\": false,\n" +
                    "    \"total_count\": 22,\n" +
                    "    \"url\": \"http://google.com\"\n" +
                    "}";

    @Test
    public void fromString_whenStringIsNull_returnsNull() {
        assertNull(Customer.fromString(null));
    }

    @Test
    public void fromJson_whenNotACustomer_returnsNull() {
        assertNull(Customer.fromString(NON_CUSTOMER_OBJECT));
    }

    @Test
    public void fromJson_whenCustomer_returnsExpectedCustomer() {
        final Customer customer = CustomerFixtures.CUSTOMER;
        assertNotNull(customer);
        assertEquals("cus_AQsHpvKfKwJDrF", customer.getId());
        assertEquals("abc123", customer.getDefaultSource());
        assertNull(customer.getShippingInformation());
        assertNotNull(customer.getSources());
        assertEquals("/v1/customers/cus_AQsHpvKfKwJDrF/sources", customer.getUrl());
        assertEquals(Boolean.FALSE, customer.getHasMore());
        assertEquals(Integer.valueOf(0), customer.getTotalCount());
    }

    @Test
    public void fromJson_whenCustomerHasApplePay_returnsCustomerWithoutApplePaySources()
            throws JSONException {
        final Customer customer = Customer.fromString(createTestCustomerObjectWithApplePaySource());
        assertNotNull(customer);
        assertEquals(2, customer.getSources().size());
        // Note that filtering the apple_pay sources intentionally does not change the total
        // count value.
        assertEquals(Integer.valueOf(5), customer.getTotalCount());
    }

    @Test
    public void fromJson_createsSameObject() {
        final Customer expectedCustomer = Customer.fromJson(CustomerFixtures.CUSTOMER_JSON);
        assertNotNull(expectedCustomer);
        assertEquals(
                expectedCustomer,
                Customer.fromJson(CustomerFixtures.CUSTOMER_JSON)
        );
    }

    @NonNull
    private String createTestCustomerObjectWithApplePaySource() throws JSONException {
        final JSONObject rawJsonCustomer = CustomerFixtures.CUSTOMER_JSON;
        final JSONObject sourcesObject = rawJsonCustomer.getJSONObject("sources");
        final JSONArray sourcesArray = sourcesObject.getJSONArray("data");

        sourcesObject.put("total_count", 5);
        sourcesArray.put(new JSONObject(JSON_APPLE_PAY_CARD));

        final JSONObject manipulatedCard = new JSONObject(JSON_CARD_USD);
        manipulatedCard.put("id", "card_id55555");
        manipulatedCard.put("tokenization_method", "apple_pay");

        // Note that we don't yet explicitly support bitcoin sources, but this data is
        // convenient for the test because it is not an apple pay source.
        sourcesArray.put(new JSONObject(EXAMPLE_JSON_SOURCE_WITHOUT_NULLS));
        sourcesArray.put(new JSONObject(EXAMPLE_ALIPAY_SOURCE));
        sourcesArray.put(new JSONObject(JSON_CARD_USD));
        sourcesArray.put(manipulatedCard);
        sourcesObject.put("data", sourcesArray);

        rawJsonCustomer.put("sources", sourcesObject);

        // Verify JSON manipulation
        assertTrue(rawJsonCustomer.has("sources"));
        assertTrue(rawJsonCustomer.getJSONObject("sources").has("data"));
        assertEquals(5,
                rawJsonCustomer.getJSONObject("sources").getJSONArray("data").length());

        return rawJsonCustomer.toString();
    }
}

File: stripe/src/test/java/com/stripe/android/model/CustomerTest.kt
package com.stripe.android.model

import com.stripe.android.model.CardTest.Companion.JSON_CARD_USD
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.json.JSONException
import org.json.JSONObject

/**
 * Test class for [Customer] model object.
 */
class CustomerTest {

    @Test
    fun fromString_whenStringIsNull_returnsNull() {
        assertNull(Customer.fromString(null))
    }

    @Test
    fun fromJson_whenNotACustomer_returnsNull() {
        assertNull(Customer.fromJson(NON_CUSTOMER_OBJECT))
    }

    @Test
    fun fromJson_whenCustomer_returnsExpectedCustomer() {
        val customer = CustomerFixtures.CUSTOMER
        assertNotNull(customer)
        assertEquals("cus_AQsHpvKfKwJDrF", customer.id)
        assertEquals("abc123", customer.defaultSource)
        assertNull(customer.shippingInformation)
        assertNotNull(customer.sources)
        assertEquals("/v1/customers/cus_AQsHpvKfKwJDrF/sources", customer.url)
        assertFalse(customer.hasMore)
        assertEquals(0, customer.totalCount)
    }

    @Test
    @Throws(JSONException::class)
    fun fromJson_whenCustomerHasApplePay_returnsCustomerWithoutApplePaySources() {
        val customer = Customer.fromString(createTestCustomerObjectWithApplePaySource())
        assertNotNull(customer)
        assertEquals(2, customer.sources.size)
        // Note that filtering the apple_pay sources intentionally does not change the total
        // count value.
        assertEquals(5, customer.totalCount)
    }

    @Test
    fun fromJson_createsSameObject() {
        val expectedCustomer = Customer.fromJson(CustomerFixtures.CUSTOMER_JSON)
        assertNotNull(expectedCustomer)
        assertEquals(
            expectedCustomer,
            Customer.fromJson(CustomerFixtures.CUSTOMER_JSON)
        )
    }

    @Throws(JSONException::class)
    private fun createTestCustomerObjectWithApplePaySource(): String {
        val rawJsonCustomer = CustomerFixtures.CUSTOMER_JSON
        val sourcesObject = rawJsonCustomer.getJSONObject("sources")
        val sourcesArray = sourcesObject.getJSONArray("data")

        sourcesObject.put("total_count", 5)
        sourcesArray.put(SourceFixtures.APPLE_PAY)

        val manipulatedCard = JSONObject(JSON_CARD_USD.toString())
        manipulatedCard.put("id", "card_id55555")
        manipulatedCard.put("tokenization_method", "apple_pay")

        // Note that we don't yet explicitly support bitcoin sources, but this data is
        // convenient for the test because it is not an apple pay source.
        sourcesArray.put(SourceTest.EXAMPLE_JSON_SOURCE_WITHOUT_NULLS)
        sourcesArray.put(SourceFixtures.ALIPAY_JSON)
        sourcesArray.put(JSONObject(JSON_CARD_USD.toString()))
        sourcesArray.put(manipulatedCard)
        sourcesObject.put("data", sourcesArray)

        rawJsonCustomer.put("sources", sourcesObject)

        // Verify JSON manipulation
        assertTrue(rawJsonCustomer.has("sources"))
        assertTrue(rawJsonCustomer.getJSONObject("sources").has("data"))
        assertEquals(5,
            rawJsonCustomer.getJSONObject("sources").getJSONArray("data").length())

        return rawJsonCustomer.toString()
    }

    companion object {

        private val NON_CUSTOMER_OBJECT = JSONObject(
            """
            {
                "object": "not_a_customer",
                "has_more": false,
                "total_count": 22,
                "url": "http://google.com"
            }
            """.trimIndent()
        )
    }
}
